File: pkg/rates/rates.go
package rates

import "math"

// DiscountFactor is the interface for discount factor calculations.
// Implementations must return the discount factor v^term for a given term.
type DiscountFactor interface {
	Discount(term int) float64
}

// RateConverter performs interest rate conversions and present value calculations.
// It pre-computes a discount table for terms 0 through 100 for fast lookups.
type RateConverter struct {
	EffectiveRate float64
	discountTable []float64
}

// NewRateConverter creates a RateConverter for the given effective annual rate.
// It pre-computes discount factors v^t for t in [0, 100] where v = 1/(1+i).
func NewRateConverter(effectiveRate float64) *RateConverter {
	v := 1 / (1 + effectiveRate)
	table := make([]float64, 101)
	table[0] = 1.0
	for i := 1; i <= 100; i++ {
		table[i] = table[i-1] * v
	}
	return &RateConverter{
		EffectiveRate: effectiveRate,
		discountTable: table,
	}
}

// Discount returns the discount factor v^term.
// Uses a pre-computed table for terms 0-100, falls back to loop for larger terms.
func (r *RateConverter) Discount(term int) float64 {
	if term <= 0 {
		return 1
	}
	if term < len(r.discountTable) {
		return r.discountTable[term]
	}
	v := r.V()
	result := 1.0
	for range term {
		result *= v
	}
	return result
}

// V returns the one-period discount factor v = 1/(1+i).
func (r *RateConverter) V() float64 {
	return 1 / (1 + r.EffectiveRate)
}

// VStar returns the v-star factor v* = (1+j) * v,
// used when premiums compound at rate j while being discounted at rate i.
func (r *RateConverter) VStar(j float64) float64 {
	return (1 + j) * r.V()
}

// PresentValue returns sumAssured * v^term.
func (r *RateConverter) PresentValue(sumAssured float64, term int) float64 {
	if term <= 0 {
		return sumAssured
	}
	if term < len(r.discountTable) {
		return sumAssured * r.discountTable[term]
	}
	v := r.V()
	result := sumAssured
	for range term {
		result *= v
	}
	return result
}

// PresentValueStar returns sumAssured * (v*)^term using the v-star discount factor.
func (r *RateConverter) PresentValueStar(sumAssured float64, term int, j float64) float64 {
	if term <= 0 {
		return sumAssured
	}
	vStar := r.VStar(j)
	return sumAssured * math.Pow(vStar, float64(term))
}

// NominalToEffective converts a nominal rate compounded m times per period
// to an effective annual rate: i = (1 + im/m)^m - 1.
func NominalToEffective(im float64, m int) float64 {
	return math.Pow(1+(im/float64(m)), float64(m)) - 1
}

// EffectiveToNominal converts an effective annual rate to a nominal rate
// compounded m times per period: im = m * ((1+i)^(1/m) - 1).
func EffectiveToNominal(i float64, m int) float64 {
	return float64(m) * (math.Pow(1+i, 1/float64(m)) - 1)
}

// ForceOfInterest returns the force of interest delta = ln(1+i).
func ForceOfInterest(i float64) float64 {
	return math.Log(1 + i)
}

// InterestFromForce converts a force of interest to an effective annual rate: i = e^delta - 1.
func InterestFromForce(delta float64) float64 {
	return math.Exp(delta) - 1
}

// AnnuityCertainImmediate returns the present value of an annuity-certain-immediate:
// a_angle_n = (1 - v^n) / i.
func AnnuityCertainImmediate(i float64, n int) float64 {
	if n <= 0 || i <= 0 {
		return 0
	}
	v := 1 / (1 + i)
	return (1 - math.Pow(v, float64(n))) / i
}

// AnnuityCertainDue returns the present value of an annuity-certain-due:
// adbl_angle_n = (1 - v^n) / d where d = i/(1+i).
func AnnuityCertainDue(i float64, n int) float64 {
	if n <= 0 || i <= 0 {
		return 0
	}
	v := 1 / (1 + i)
	d := i / (1 + i)
	return (1 - math.Pow(v, float64(n))) / d
}

// MacaulayDuration computes the Macaulay duration of a cash flow stream.
// cashFlows[t] is the cash flow at time t (t=1..len(cashFlows)).
// Returns sum(t * PV_t) / sum(PV_t).
func MacaulayDuration(i float64, cashFlows []float64) float64 {
	if i <= 0 || len(cashFlows) == 0 {
		return 0
	}
	v := 1 / (1 + i)
	pvTotal := 0.0
	duration := 0.0
	for t, cf := range cashFlows {
		if cf <= 0 {
			continue
		}
		pv := cf * math.Pow(v, float64(t+1))
		pvTotal += pv
		duration += float64(t+1) * pv
	}
	if pvTotal <= 0 {
		return 0
	}
	return duration / pvTotal
}

// ModifiedDuration computes the modified duration: MacaulayDuration / (1+i).
// This measures the sensitivity of bond price to interest rate changes.
func ModifiedDuration(i float64, cashFlows []float64) float64 {
	md := MacaulayDuration(i, cashFlows)
	return md / (1 + i)
}

// Convexity computes the convexity of a cash flow stream.
// Returns sum(t*(t+1)*PV_t) / ((1+i)^2 * PV_total).
func Convexity(i float64, cashFlows []float64) float64 {
	if i <= 0 || len(cashFlows) == 0 {
		return 0
	}
	v := 1 / (1 + i)
	pvTotal := 0.0
	conv := 0.0
	for t, cf := range cashFlows {
		if cf <= 0 {
			continue
		}
		pv := cf * math.Pow(v, float64(t+1))
		pvTotal += pv
		conv += float64(t+1) * float64(t+2) * pv
	}
	if pvTotal <= 0 {
		return 0
	}
	return conv / (pvTotal * (1 + i) * (1 + i))
}

File: zig/src/rates.zig
// rates.zig - Interest rate calculations, discount factors, present value
//
// Port of Go pkg/rates. Provides pre-computed discount table for O(1) lookups
// and actuarial rate conversion functions.
const std = @import("std");
const math = std.math;

/// RateConverter performs interest rate conversions and present value calculations.
/// Pre-computes a dynamically-growing discount table for fast lookups.
pub const RateConverter = struct {
    discount_table: []f64,
    effective_rate: f64,
    allocator: std.mem.Allocator,

    /// Creates a RateConverter for the given effective annual rate.
    /// Pre-computes discount factors v^t for t in [0, 100] where v = 1/(1+i).
    pub fn init(allocator: std.mem.Allocator, effective_rate: f64) !RateConverter {
        const v_factor = 1.0 / (1.0 + effective_rate);
        const table = try allocator.alloc(f64, 101);
        table[0] = 1.0;
        for (1..101) |i| {
            table[i] = table[i - 1] * v_factor;
        }
        return .{
            .discount_table = table,
            .effective_rate = effective_rate,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *RateConverter) void {
        self.allocator.free(self.discount_table);
    }

    /// Returns the discount factor v^term.
    /// Uses pre-computed table that grows as needed.
    pub fn discount(self: *RateConverter, term: i64) f64 {
        if (term <= 0) return 1.0;
        const t: usize = @intCast(term);
        if (t >= self.discount_table.len) {
            self.growTable(t);
        }
        return self.discount_table[t];
    }

    fn growTable(self: *RateConverter, min_size: usize) void {
        const new_size: usize = @max(min_size + 1, self.discount_table.len * 2 + 1);
        const new_table = self.allocator.realloc(self.discount_table, new_size) catch {
            // If realloc fails, just recompute into a new allocation
            const buf = self.allocator.alloc(f64, new_size) catch return;
            self.allocator.free(self.discount_table);
            self.discount_table = buf;
            self.recomputeTable(new_size);
            return;
        };
        self.discount_table = new_table;
        self.recomputeTable(new_size);
    }

    fn recomputeTable(self: *RateConverter, size: usize) void {
        const v_factor = 1.0 / (1.0 + self.effective_rate);
        self.discount_table[0] = 1.0;
        for (1..size) |i| {
            self.discount_table[i] = self.discount_table[i - 1] * v_factor;
        }
    }

    /// Returns the one-period discount factor v = 1/(1+i).
    pub fn v(self: *const RateConverter) f64 {
        return 1.0 / (1.0 + self.effective_rate);
    }

    /// Returns the v-star factor v* = (1+j) * v,
    /// used when premiums compound at rate j while being discounted at rate i.
    pub fn vStar(self: *const RateConverter, j: f64) f64 {
        return (1.0 + j) * self.v();
    }

    /// Returns sumAssured * v^term.
    pub fn presentValue(self: *RateConverter, sum_assured: f64, term: i64) f64 {
        if (term <= 0) return sum_assured;
        const t: usize = @intCast(term);
        if (t >= self.discount_table.len) {
            self.growTable(t);
        }
        return sum_assured * self.discount_table[t];
    }

    /// Returns sumAssured * (v*)^term using the v-star discount factor.
    pub fn presentValueStar(self: *const RateConverter, sum_assured: f64, term: i64, j: f64) f64 {
        if (term <= 0) return sum_assured;
        return sum_assured * math.pow(f64, self.vStar(j), @floatFromInt(term));
    }
};

// ---------------------------------------------------------------------------
// Rate conversion functions
// ---------------------------------------------------------------------------

/// Converts a nominal rate compounded m times per period to effective annual rate.
/// i = (1 + im/m)^m - 1
pub fn nominalToEffective(im: f64, m: i64) f64 {
    const mf: f64 = @floatFromInt(m);
    return math.pow(f64, 1.0 + im / mf, mf) - 1.0;
}

/// Converts an effective annual rate to nominal rate compounded m times per period.
/// im = m * ((1+i)^(1/m) - 1)
pub fn effectiveToNominal(i: f64, m: i64) f64 {
    const mf: f64 = @floatFromInt(m);
    return mf * (math.pow(f64, 1.0 + i, 1.0 / mf) - 1.0);
}

/// Returns the force of interest delta = ln(1+i).
pub fn forceOfInterest(i: f64) f64 {
    return @log(1.0 + i);
}

/// Converts a force of interest to an effective annual rate: i = e^delta - 1.
pub fn interestFromForce(delta: f64) f64 {
    return @exp(delta) - 1.0;
}

/// Returns the present value of an annuity-certain-immediate: a_angle_n = (1 - v^n) / i.
pub fn annuityCertainImmediate(i: f64, n: i64) f64 {
    if (n <= 0 or i <= 0) return 0.0;
    const v = 1.0 / (1.0 + i);
    return (1.0 - math.pow(f64, v, @floatFromInt(n))) / i;
}

/// Returns the present value of an annuity-certain-due: adbl_angle_n = (1 - v^n) / d.
pub fn annuityCertainDue(i: f64, n: i64) f64 {
    if (n <= 0 or i <= 0) return 0.0;
    const v = 1.0 / (1.0 + i);
    const d = i / (1.0 + i);
    return (1.0 - math.pow(f64, v, @floatFromInt(n))) / d;
}

/// Macaulay duration of a cash flow stream.
/// Returns sum(t * PV_t) / sum(PV_t).
pub fn macaulayDuration(i: f64, cash_flows: []const f64) f64 {
    if (i <= 0 or cash_flows.len == 0) return 0.0;
    const v = 1.0 / (1.0 + i);
    var v_pow = v;
    var pv_total: f64 = 0.0;
    var duration: f64 = 0.0;
    for (cash_flows, 0..) |cf, t| {
        if (cf <= 0) {
            v_pow *= v;
            continue;
        }
        const pv = cf * v_pow;
        pv_total += pv;
        duration += @as(f64, @floatFromInt(t + 1)) * pv;
        v_pow *= v;
    }
    if (pv_total <= 0) return 0.0;
    return duration / pv_total;
}

/// Modified duration: MacaulayDuration / (1+i).
pub fn modifiedDuration(i: f64, cash_flows: []const f64) f64 {
    return macaulayDuration(i, cash_flows) / (1.0 + i);
}

/// Convexity of a cash flow stream.
pub fn convexity(i: f64, cash_flows: []const f64) f64 {
    if (i <= 0 or cash_flows.len == 0) return 0.0;
    const v = 1.0 / (1.0 + i);
    var v_pow = v;
    var pv_total: f64 = 0.0;
    var conv: f64 = 0.0;
    for (cash_flows, 0..) |cf, t| {
        if (cf <= 0) {
            v_pow *= v;
            continue;
        }
        const pv = cf * v_pow;
        pv_total += pv;
        conv += @as(f64, @floatFromInt(t + 1)) * @as(f64, @floatFromInt(t + 2)) * pv;
        v_pow *= v;
    }
    if (pv_total <= 0) return 0.0;
    return conv / (pv_total * (1.0 + i) * (1.0 + i));
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "NominalToEffective" {
    const testing = std.testing;
    try testing.expectApproxEqAbs(@as(f64, 0.05), nominalToEffective(0.05, 1), 1e-10);
    // Monthly compounding of 12%
    const me = nominalToEffective(0.12, 12);
    try testing.expect(me > 0.12);
    try testing.expect(me < 0.13);
}

test "EffectiveToNominal" {
    const testing = std.testing;
    try testing.expectApproxEqAbs(@as(f64, 0.05), effectiveToNominal(0.05, 1), 1e-10);
}

test "ForceOfInterest" {
    const testing = std.testing;
    const delta = forceOfInterest(0.05);
    try testing.expectApproxEqAbs(@as(f64, 0.04879), delta, 1e-3);
    try testing.expectApproxEqAbs(@as(f64, 0.05), interestFromForce(delta), 1e-10);
}

test "AnnuityCertain" {
    const testing = std.testing;
    const a_imd = annuityCertainImmediate(0.05, 10);
    try testing.expect(a_imd > 0);
    try testing.expect(a_imd < 10.0);

    const a_dbl = annuityCertainDue(0.05, 10);
    try testing.expect(a_dbl > a_imd);
}

test "RateConverter basics" {
    const testing = std.testing;
    var rc = try RateConverter.init(testing.allocator, 0.05);
    defer rc.deinit();

    try testing.expectApproxEqAbs(@as(f64, 1.0 / 1.05), rc.v(), 1e-10);
    try testing.expectApproxEqAbs(@as(f64, 1.0), rc.discount(0), 1e-10);
    try testing.expectApproxEqAbs(@as(f64, 1.0 / 1.05), rc.discount(1), 1e-10);

    // Present value
    const pv = rc.presentValue(1000, 1);
    try testing.expectApproxEqAbs(@as(f64, 952.38), pv, 0.01);

    // v-star
    const vs = rc.vStar(0.02);
    try testing.expectApproxEqAbs(@as(f64, 1.02 / 1.05), vs, 1e-10);
}

test "PresentValue" {
    const testing = std.testing;
    var rc = try RateConverter.init(testing.allocator, 0.05);
    defer rc.deinit();

    try testing.expectApproxEqAbs(@as(f64, 1000.0), rc.presentValue(1000, 0), 0.01);
    try testing.expectApproxEqAbs(@as(f64, 952.38), rc.presentValue(1000, 1), 0.01);
    try testing.expectApproxEqAbs(@as(f64, 783.53), rc.presentValue(1000, 5), 0.01);
}

test "PresentValueStar" {
    const testing = std.testing;
    var rc = try RateConverter.init(testing.allocator, 0.05);
    defer rc.deinit();
    const j: f64 = 0.02;

    try testing.expectApproxEqAbs(@as(f64, 1000.0), rc.presentValueStar(1000, 0, j), 0.01);
    try testing.expectApproxEqAbs(@as(f64, 971.43), rc.presentValueStar(1000, 1, j), 0.01);
    try testing.expectApproxEqAbs(@as(f64, 865.08), rc.presentValueStar(1000, 5, j), 0.01);
}

test "Macaulay duration" {
    const testing = std.testing;
    const cash_flows = [_]f64{ 100, 100, 100, 100, 1100 };
    const md = macaulayDuration(0.05, &cash_flows);
    try testing.expect(md > 4.0);
    try testing.expect(md < 5.0);
}
